mithril-aggregator/src/services/signed_entity.rs

```diff
@@ -165,44 +215,41 @@ impl SignedEntityService for MithrilSignedEntityService {
         &self,
         signed_entity_type: SignedEntityType,
         certificate: &Certificate,
-    ) -> StdResult<()> {
-        info!(
-            "MithrilSignedEntityService::create_artifact";
-            "signed_entity_type" => ?signed_entity_type,
-            "certificate_hash" => &certificate.hash
-        );
-
-        let mut remaining_retries = 2;
-        let artifact = loop {
-            remaining_retries -= 1;
-
-            match self
-                .compute_artifact(signed_entity_type.clone(), certificate)
-                .await
-            {
-                Err(error) if remaining_retries == 0 => break Err(error),
-                Err(_error) => (),
-                Ok(artifact) => break Ok(artifact),
-            };
-        }?;
-
-        let signed_entity = SignedEntityRecord {
-            signed_entity_id: artifact.get_id(),
-            signed_entity_type: signed_entity_type.clone(),
-            certificate_id: certificate.hash.clone(),
-            artifact: serde_json::to_string(&artifact)?,
-            created_at: Utc::now(),
-        };
-
-        self.signed_entity_storer
-            .store_signed_entity(&signed_entity)
+    ) -> StdResult<JoinHandle<StdResult<()>>> {
+        if self
+            .signed_entity_type_lock
+            .is_locked(&signed_entity_type)
             .await
-            .with_context(|| {
-                format!(
-                    "Signed Entity Service can not store signed entity with type: '{signed_entity_type}'"
-                )
-            })?;
-        Ok(())
+        {
+            return Err(anyhow!(
+                "Signed entity type '{:?}' is already locked",
+                signed_entity_type
+            ));
+        }
+
+        let service = self.clone();
+        let certificate_cloned = certificate.clone();
+        service
+            .signed_entity_type_lock
+            .lock(&signed_entity_type)
+            .await;
+
+        Ok(tokio::task::spawn(async move {
+            let signed_entity_type_clone = signed_entity_type.clone();
+            let service_clone = service.clone();
+            let result = tokio::task::spawn(async move {
+                service_clone
+                    .create_artifact_task(signed_entity_type_clone, &certificate_cloned)
+                    .await
+            })
+            .await;
+            service
+                .signed_entity_type_lock
+                .release(signed_entity_type)
+                .await;
+
+            result.unwrap()
```

Comment:
Can we avoid an `unwrap` here?

Reply:
It will be replaced by a `result.with_context